src/Ormolu/Processing/Preprocess.hs

```diff
@@ -29,79 +31,89 @@ preprocess ::
   String ->
   -- | Region deltas
   RegionDeltas ->
-  -- | Literal prefix, pre-processed input, literal suffix, extra comments
-  (String, String, String, [Located String])
+  -- | Literal prefix, pre-processed input, literal suffix, extra comments,
+  -- disabled regions
+  (String, String, String, [Located String], DisabledRegions)
 preprocess path input RegionDeltas {..} =
-  go 1 OrmoluEnabled Cpp.Outside id id regionLines
+  go 1 Cpp.Outside id id enabledLines
   where
     (prefixLines, otherLines) = splitAt regionPrefixLength (lines input)
     (regionLines, suffixLines) =
       let regionLength = length otherLines - regionSuffixLength
        in splitAt regionLength otherLines
-    go !n ormoluState cppState inputSoFar csSoFar = \case
+    (disabledRegions, enabledLines) = cutDisabledRegions regionLines
+    go !n cppState inputSoFar csSoFar = \case
       [] ->
-        let input' = unlines (inputSoFar [])
-         in ( unlines prefixLines,
-              case ormoluState of
-                OrmoluEnabled -> input'
-                OrmoluDisabled -> input' ++ endDisabling,
-              unlines suffixLines,
-              csSoFar []
-            )
+        ( unlines prefixLines,
+          unlines (inputSoFar []),
+          unlines suffixLines,
+          csSoFar [],
+          disabledRegions
+        )
       (x : xs) ->
-        let (x', ormoluState', cppState', cs) =
-              processLine path n ormoluState cppState x
+        let (x', cppState', cs) =
+              processLine path n cppState x
          in go
               (n + 1)
-              ormoluState'
               cppState'
               (inputSoFar . (x' :))
               (csSoFar . (maybeToList cs ++))
               xs
 
+-- | Cut regions where formatting is disabled, replacing each of them with
+-- a magic comment.
+cutDisabledRegions ::
+  -- | Lines to process
+  [String] ->
+  -- | Regions where formatting is disabled, all remaining lines
+  (DisabledRegions, [String])
+cutDisabledRegions = go (id, id)
+  where
+    cut xs = (prefix, disabledRegionLines, suffix)
+      where
+        (prefix, rest) = break isOrmoluDisable xs
+        (disabledRegionLines, suffix) = break isOrmoluEnable rest
+    removeIndentation xs = drop (getIndentationLevel 0 xs) <$> xs
+    go (regionsSoFar, linesSoFar) (cut -> (pre, [], _)) =
```

Comment:
The problem with this is that GHC will see an input with a different number of lines in total and therefore its error messages (when the parser fails) will be shifted and inaccurate. Consider this:

```haskell
module Main (main) where

{- ORMOLU_DISABLE -}
foo Int
foo Int
foo Int
foo Int
foo Int
foo Int
foo Int
foo 5 EU>>>>

{- ORMOLU_ENABLE -}

main :: IO ()
main = return <<<
```

The last line fails to parse. On `master` it report the problematic location correctly:

```
disabled-test.hs:17:1
  The GHC parser (in Haddock mode) failed:
  parse error (possibly incorrect indentation or mismatched brackets)
```

Note line 17.

With your change the region in the middle is cut, so the error messages becomes misleading:

```
disabled-test.hs:8:1
  The GHC parser (in Haddock mode) failed:
  parse error (possibly incorrect indentation or mismatched brackets)
```

Note line 8.

This is, in fact, was the main motivation for doing things the way it is done right now. Fortunately, I think we can work this around if we insert the right number of blank lines between `disableMarker` and `enableMarker`.

Reply:
Thanks! I've tried the idea with blank lines and it seems to work. I'm going for holidays for now, will get back to this mid-July.

Reply:
Looks like the blank lines work. Significant changes in the last commit:

* Replacing disabled regions with blank lines.
* Normalize all `{- ORMOLU_ENABLE -}` magic comments (in the previous version only the comments at the beginning and the end of a disabled region were normalized, which differs from `master`). I decided not to normalize the `{- ORMOLU_DISABLE -}` comments inside disabled regions. See `data/examples/other/disabling/normalize-magic-comments.hs`.
* `Ormolu.Utils.getIndentationLevel` was made total (with the provided default value used for empty lists, which is not very clean).
* Rewording of `cutDisabledRegions` and `pasteDisabledRegions`. I find the new code somewhat more readable, thought I'm not really satisfied with it.

However, this is still work in progress. I'm now afraid that moving the processing of disabled regions out of `processLine` was a wrong decision, as it may cause problems in interplay between disabled regions and CPP. I'll have to check this, but it's late today and, once again, I won't have time for it for a few days.

Perhaps I'll also try to include your `disabled-test.hs` as a test case.

Reply:
It turns out that mixing magic comments with CPP or LINE pragmas results in a mess, especially when one disables formatting in a fragment of a definition (which is not much of a surprise). In particular running the `master` version of  `ormolu` on any of the following files result in the "AST of input and AST of formatted code differ" error, and the `--unsafe` output doesn't look very good.

```haskell
f =
  g
    {- ORMOLU_DISABLE -}
#ifdef ABC
    h
#else
    i
#endif
    j
    {- ORMOLU_ENABLE -}
    k
```

```haskell
f =
  g
    {- ORMOLU_DISABLE -}
    h
    i
{-# LINE 42 "Foo.vhs" #-}
    j
    {- ORMOLU_ENABLE -}
    k
```

Unfortunately keeping backward compatibility with the present behaviour and improving it are both somewhat tricky. Perhaps the latter may  be easier.